### PR TITLE
feat(mqttbr): describe MQTT bridge pools

### PR DESCRIPTION
**Pull request by keynslug**
* New configuration options.
* Format of Client IDs.
* Consequences of using non-shared subscription topics.

### CODE REVIEW

File: en_US/data-integration/data-bridge-mqtt.md [hunk 1]
```diff
@@ -73,24 +73,29 @@ The following section will use EMQX [public MQTT broker](https://www.emqx.com/en
...
-       - QoS: Select `0`.
+       - **QoS**: Select `0`.
      
      - **Local MQTT Broker**: Forward the received messages to specific local topics or leave them blank, then these messages will first be processed by the configured rules and then forwarded with the [republish action](./rules.md).
        - **Topic**: Input `local/topic/ingress`.
        - **QoS**: Select `0` or `${qos}` (to use the QoS of the received messages).
        - **Retain**: Confirm whether the message will be published as a retained message.
        - **Payload**: Payload template for the messages to be forwarded, and Supports reading data using `${field}` syntax.
+
+     - **Pool size**: Specify the size of the pool of MQTT clients that will ingest messages from the remote broker, we'll set `8` to make use of available parallelism. This is safe as long as shared subscription is used for the remote topic.
```
Comment (zmstone) [inline en_US/data-integration/data-bridge-mqtt.md]:
```suggestion
     - **Pool size**: Introduced since 5.1, this config is to specify the size of the pool of MQTT clients that will ingest messages from the remote broker, we'll set `8` to make use of available parallelism. This is safe as long as shared subscription is used for the remote topic.
```

File: en_US/data-integration/data-bridge-mqtt.md [hunk 2]
```diff
@@ -73,24 +73,29 @@ The following section will use EMQX [public MQTT broker](https://www.emqx.com/en
...
        - **Topic**: Input `local/topic/egress`.
+
      - **Remote MQTT Broker**: Specify the target topics on the remote broker.
        - **Topic**: Input `remote/topic/egress`.
        - **QoS**: Select `0` or `${qos}` (to use the QoS of the received messages).
        - **Retain**: Confirm whether the message will be published as a retained message.
        - **Payload**: Payload template for the messages to be forwarded, and Supports reading data using `${field}` syntax.
-   
+
+     - **Pool size**: Specify the size of the pool of MQTT clients that will publish messages to the remote broker, we'll set `8` to make use of available parallelism.
```
Comment (zmstone) [inline en_US/data-integration/data-bridge-mqtt.md]:
```suggestion
     - **Pool size**: Introduced in 5.1, this config is to specify the size of the pool of MQTT clients that will publish messages to the remote broker, we'll set `8` to make use of available parallelism.
```

File: en_US/data-integration/data-bridge-mqtt.md [hunk 3]
```diff
@@ -127,9 +133,31 @@ bridges.mqtt.my_mqtt_bridge {
...
+[${ClientIDPrefix}:]${BridgeName}:${Mode}:${NodeName}:${N}
+```
+
+| Fragment             | Description                     |
+| -------------------- | ------------------------------- |
+| `${ClientIDPrefix}`  | As per configuration. If not set then whole first fragment is omitted. |
+| `${BridgeName}`      | User-provided name of the bridge. |
+| `${Mode}`            | Either `ingress` or `egress`. |
+| `${NodeName}`        | [Name of the node](../../configuration/cluster.html#node-names) on which the MQTT client is running. |
+| `${N}`               | Number between 1 and the configured size of the MQTT client pool. |
```
Comment (zmstone) [inline en_US/data-integration/data-bridge-mqtt.md]:
```suggestion
| `${N}`               | Number from `1` to the configured size of the MQTT client pool. |
```

**Review by zmstone — DISMISSED** [review]

File: en_US/data-integration/data-bridge-mqtt.md [hunk 4]
```diff
@@ -127,9 +133,31 @@ bridges.mqtt.my_mqtt_bridge {
...
+```bash
+[${ClientIDPrefix}:]${BridgeName}:${Mode}:${NodeName}:${N}
+```
+
+| Fragment             | Description                     |
+| -------------------- | ------------------------------- |
+| `${ClientIDPrefix}`  | As per configuration. If not set then whole first fragment is omitted. |
+| `${BridgeName}`      | User-provided name of the bridge. |
+| `${Mode}`            | Either `ingress` or `egress`. |
+| `${NodeName}`        | [Name of the node](../../configuration/cluster.html#node-names) on which the MQTT client is running. |
```
Comment (wivwiv) [inline en_US/data-integration/data-bridge-mqtt.md]:
```suggestion
| `${NodeName}`        | [Name of the node](../../configuration/cluster.md#node-names) on which the MQTT client is running. |
```

**Review by wivwiv — CHANGES_REQUESTED** [review]

**Comment by wivwiv** [comment]
@Meggielqk Please translate to zh

**Review by Meggielqk — COMMENTED** [review]
中文文件里 “MQTT 代理” 是否可以都改为“MQTT 服务器”/“MQTT 服务”？ @wivwiv

File: en_US/data-integration/data-bridge-mqtt.md [hunk 5]
```diff
@@ -1,56 +1,71 @@
...
+Similar to the ingress mode, the MQTT data bridge can also be used in conjunction with rules. In **egress** direction, the data bridge can be used as the action of the rule:
 
-The message flow for MQTT data bridge working with rules are as follows:
+<img src="./assets/bridge_egress_rule.png" alt="bridge_egress_rule" style="zoom: 50%;" />
 
-In **ingress** direction, it can be used as the data source of the rule:
+### Connection Pool
 
-<img src="./assets/bridge_igress_rule_link.png" alt="bridge_igress_rule_link" style="zoom:50%;" />
+EMQX allows more than one client to connect to the bridged MQTT broker at the same time. When creating the data bridge, you can set a pool of MQTT client connetions and configure the pool size indicating the number of the client connections in the pool. Enabling connection pool in MQTT data bridge allows for efficient utilization of server resources, leading to greater message throughput and improved concurrency performance. This brings benefits in scenarios with high loads and concurrent connections.
```
Comment (keynslug) [inline en_US/data-integration/data-bridge-mqtt.md]:
```suggestion
EMQX allows more than one client to connect to the bridged MQTT broker at the same time. When creating the data bridge, you can set a pool of MQTT client connections and configure the pool size indicating the number of the client connections in the pool. Enabling connection pool allows for efficient utilization of server resources, leading to greater message throughput.
```

IMO there is unnecessary repetition.

File: en_US/data-integration/data-bridge-mqtt.md [hunk 6]
```diff
@@ -1,56 +1,71 @@
...
+| `${BridgeName}`     | User-provided name of the bridge.                            |
+| `${Mode}`           | Either `ingress` or `egress`.                                |
+| `${NodeName}`       | [Name of the node](../configuration/cluster.md#node-names) on which the MQTT client is running. |
+| `${N}`              | Number from `1` to the configured size of the connection pool. |
+
+#### Use Connection Pool in Ingress Mode
+
+Although the connection pool applies to both ingress and egress modes, using the connection pool in ingress mode has more requirements. When data bridges in ingress mode are created between MQTT brokers in different MQTT [clusters](../deploy/cluster/introduction.md), clients in the connection pool will receive duplicated messages from the remote broker if they all subscribe to the same topic. In that case, it will bring pressure to the brokers, so it is strongly advised to use [shared subscription](../messaging/mqtt-shared-subscription.md) as a kind of safety measure. For example, you can configure the topic of the remote MQTT Broker to  `$share/name1/topic1` or `$share/name2/topic2/#` if topic filter is used.
```
Comment (keynslug) [inline en_US/data-integration/data-bridge-mqtt.md]: > requirements

I'd say this is a _caveat_ rather than a _requirement_.

Also, this paragraph kinda lost large part of its initial message. There's no mention anymore that in non-shared subscription case pool downscales to a single client.

> When data bridges in ingress mode are created between MQTT brokers in different MQTT [clusters](../deploy/cluster/introduction.md)

This sounds a bit misleading. Issue arises when user has EMQX cluster with more than 1 node, and configures an ingress MQTT bridge with non-shared topic to subscribe to on the remote broker. The remote broker does not have to be EMQX, it can be anything.

Bottom line is that paragraph was initially about when and how user expectations about _bridging_ will **break** under different configurations (EMQX cluster / single node, shared / non-shared subscription), namely message duplication.

File: en_US/data-integration/data-bridge-mqtt.md [hunk 7]
```diff
@@ -1,56 +1,71 @@
...
+<img src="./assets/bridge_egress_rule.png" alt="bridge_egress_rule" style="zoom: 50%;" />
 
-In **ingress** direction, it can be used as the data source of the rule:
+### Connection Pool
 
-<img src="./assets/bridge_igress_rule_link.png" alt="bridge_igress_rule_link" style="zoom:50%;" />
+EMQX allows more than one client to connect to the bridged MQTT broker at the same time. When creating the data bridge, you can set a pool of MQTT client connetions and configure the pool size indicating the number of the client connections in the pool. Enabling connection pool in MQTT data bridge allows for efficient utilization of server resources, leading to greater message throughput and improved concurrency performance. This brings benefits in scenarios with high loads and concurrent connections.
 
-In **egress** direction, it can be used as the action of the rule:
+Because the MQTT protocol requires that a client connects to one broker must has it's own unique client ID, each client in the connection pool is assigned to a unique client ID. To make the client ID predictable, EMQX automatically generates the Client ID following the pattern below:
```
Comment (keynslug) [inline en_US/data-integration/data-bridge-mqtt.md]:
```suggestion
Because the MQTT protocol requires that a client connecting to the broker must have its own unique Client ID, EMQX automatically generates the Client ID for each client in the connection pool following the pattern below:
```

File: en_US/data-integration/data-bridge-mqtt.md (hunk 6 above)
Comment (Meggielqk) [reply en_US/data-integration/data-bridge-mqtt.md]: Thank you for your comments. I have modified the descriptions.

**Review by wivwiv — DISMISSED** [review]

File: en_US/data-integration/data-bridge-mqtt.md [hunk 8]
```diff
@@ -46,7 +46,7 @@ Because the MQTT protocol requires that a client connects to one broker must has
...
+Although the connection pool applies to both ingress and egress modes, using the connection pool in ingress mode should consider a few caveats. When you have EMQX [cluster](../deploy/cluster/introduction.md) with more than 1 node, and configures an ingress MQTT bridge to subscribe non-shared topics from the remote broker, clients in the connection pool will receive duplicated messages if they all subscribe to the same topic. In that case, it will bring pressure to the brokers, so it is strongly advised to use [shared subscription](../messaging/mqtt-shared-subscription.md) as a kind of safety measure. For example, you can configure the topic of the remote MQTT Broker to `$share/name1/topic1` or `$share/name2/topic2/#` if topic filter is used. In non-shared subscription cases, the MQTT client connection pool will be downscaled to one client, which means only one client will be started.
```
Comment (keynslug) [inline en_US/data-integration/data-bridge-mqtt.md]:
```suggestion
Although the connection pool applies to both ingress and egress modes, when using the connection pool in ingress mode you should consider a few caveats. When you have EMQX [cluster](../deploy/cluster/introduction.md) with more than 1 node, and configure an ingress MQTT bridge to subscribe non-shared topics from the remote broker, clients in the connection pool will receive duplicated messages since they all subscribe to the same topic. In that case, it will bring pressure to the brokers, so it is strongly advised to use [shared subscription](../messaging/mqtt-shared-subscription.md) as a kind of safety measure. For example, you can configure the topic of the remote MQTT Broker to `$share/name1/topic1` or `$share/name2/topic2/#` if topic filter is used. In non-shared subscription cases, the MQTT client connection pool will be downscaled to one client, which means only one client will be started.
```

**Review by wivwiv — APPROVED** [review]
